featurebyte/query_graph/sql/aggregator/forward.py

```diff
@@ -0,0 +1,170 @@
+        """
+        Get aggregation subquery that performs the forward aggregation. The list of aggregation
+        specifications provided can be done in a single groupby operation.
+
+        Parameters
+        ----------
+        specs: list[ForwardAggregateSpec]
+            Aggregation specifications
+
+        Returns
+        -------
+        LeftJoinableSubquery
+        """
+        spec = specs[0]
+
+        # End point expression
+        point_in_time_expr = get_qualified_column_identifier(SpecialColumnName.POINT_IN_TIME, "REQ")
+        point_in_time_epoch_expr = self.adapter.to_epoch_seconds(point_in_time_expr)
+        horizon_in_seconds = 0
+        if spec.parameters.horizon:
+            horizon_in_seconds = pd.Timedelta(spec.parameters.horizon).total_seconds()
+        end_point_expr: Expression = cast(
+            Expression,
+            parse_one(f"FLOOR({point_in_time_epoch_expr.sql()} + {horizon_in_seconds})"),
+        )
+
+        # Get valid records (timestamp column is within the point in time, and point in time + horizon)
+        # TODO: update to range join
+        table_timestamp_col = get_qualified_column_identifier(
+            spec.parameters.timestamp_col, "TABLE"
+        )
+        # Convert to epoch seconds
+        table_timestamp_col = self.adapter.to_epoch_seconds(table_timestamp_col)
+        record_validity_condition = expressions.and_(
+            expressions.GT(
+                this=table_timestamp_col,
+                expression=point_in_time_expr,
+            ),
+            expressions.LTE(
+                this=table_timestamp_col,
+                expression=end_point_expr,
+            ),
+        )
+
+        # Join the valid records, and the request table
+        groupby_keys = [
+            GroupbyKey(
+                expr=point_in_time_expr,
+                name=SpecialColumnName.POINT_IN_TIME,
+            )
+        ] + [
+            GroupbyKey(
+                expr=get_qualified_column_identifier(serving_name, "REQ"),
+                name=serving_name,
+            )
+            for serving_name in spec.serving_names
+        ]
+        value_by = (
+            GroupbyKey(
+                expr=get_qualified_column_identifier(spec.parameters.value_by, "TABLE"),
+                name=spec.parameters.value_by,
+            )
+            if spec.parameters.value_by
+            else None
+        )
+        groupby_columns = [
+            GroupbyColumn(
+                agg_func=s.parameters.agg_func,
+                parent_expr=(
+                    get_qualified_column_identifier(s.parameters.parent, "TABLE")
+                    if s.parameters.parent
+                    else None
+                ),
+                result_name=s.agg_result_name,
+            )
+            for s in specs
+        ]
+        serving_name = quoted_identifier(spec.serving_names[0]).sql()
+        join_table_key = quoted_identifier(spec.parameters.keys[0]).sql()
```

Comment:
There could be multiple join keys and in that case they should be `and_`'ed.

Reply:
hm i assumed that targets will only have one serving name / join key. is that different?

Reply:
In the groupby method the join key can be a list. The forward aggregate is called through the same groupby method right?

Reply:
oh yes, sorry i got confused. i'll fix this in a follow-up since the tests here are almost done 😁